Service Operator: ServiceEntry dynamic creation for S3 and RDS

Krenair opened the pull request:
Particularly S3 due to trouble in https://github.com/alphagov/verify-cluster-config/pull/62

Krenair committed:
Service Operator: Automatically provision a ServiceEntry where relevant

This is for things we create that have custom hostnames, ports, etc. as these
can be tricky to set up egress whitelisting for on a cluster-level basis.

Krenair committed:
This is probably necessary

Krenair committed:
Move horrible istio schema registration stuff to a separate package

chrisfarms commented on components/service-operator/apis/storage/v1beta1/s3_types_test.go:
```diff
@@ -49,6 +50,28 @@ var _ = Describe("S3Bucket", func() {
+		Expect(err).NotTo(HaveOccurred())
+		Expect(ret.GetObjectMeta().Name).To(Equal(fmt.Sprintf("svcop-s3-%s", o.GetName())))
+		Expect(ret.GetObjectMeta().Namespace).To(Equal(o.GetNamespace()))
+		Expect(ret.GetSpec()["resolution"]).To(Equal("DNS"))
```
Could we make use of some better matchers here like `HaveKeyWithValue` so that the errors are easier to grok

chrisfarms commented on components/service-operator/apis/storage/v1beta1/s3_types_test.go:
```diff
@@ -49,6 +50,28 @@ var _ = Describe("S3Bucket", func() {
+		Expect(len(ports)).To(BeNumerically(">", 0))
+		port, ok := ports[0].(map[string]interface{})
+		Expect(ok).To(BeTrue())
+		Expect(port["name"]).To(Equal("https"))
```
Similar to above... `HaveKeyWithValue` would lead to some nicer failure messages here

chrisfarms commented on components/service-operator/internal/aws/cloudformation/controller.go:
```diff
@@ -227,9 +227,52 @@ func (r *Controller) reconcileObjectWithContext(ctx context.Context, req ctrl.Re
+	}
+	op, err := controllerutil.CreateOrUpdate(ctx, r.KubernetesClient, serviceEntry, func() error {
+		return nil
+	})
```
The serviceEntry object should be mutated inside this `func` ... for example it should be something like:

```
op, err := controllerutil.CreateOrUpdate(ctx, r.KubernetesClient, serviceEntry, func() error {
    serviceEntry.Spec = o.GetServiceEntry().Spec
    return nil
})
```

chrisfarms reviewed (dismissed):
Few comments inline...

Could we also add two tests:

* postgres_cloudformation.go should test that a ServiceEntry was created in cluster as expected with correct rds endpoint
* s3_cloudformation.go should test that a ServiceEntry was created in cluster as expected with correct bucket endpoint

Krenair committed:
Improve how I set up the tests

Krenair committed:
Entirely untested changes around ServiceEntry updating

TODO: add tests for GetServiceEntryName
TODO: try this all out in sandbox again

Krenair committed:
address Chris' comments

Krenair committed:
fmt, allow ServiceEntry name to be configured, add a couple of tests

Krenair committed:
use new configurable service-entry name

Krenair committed:
Stop prefixing ServiceEntry

Krenair committed:
Add istio ServiceEntry CRD so tests can work

Krenair committed:
Fix fake value provided by fake_client_happy to be a valid integer

FAKE_VALUE is not a valid integer and therefore cannot be a port number like
the Postgres code in here will expect. 0 is though.

chrisfarms commented:
If this is ready to go, can you tidy/fixup the commits up a bit

philandstuff commented on components/service-operator/internal/aws/cloudformation/types.go:
```diff
@@ -39,3 +39,11 @@ type StackSecretOutputter interface {
 	object.SecretNamer
 }
+
+// ServiceEntryNamer names a ServiceEntry to hold what are essentially egress
```
```suggestion
// ServiceEntryCreator names a ServiceEntry to hold what are essentially egress
```

philandstuff commented on components/service-operator/apis/storage/v1beta1/s3_types.go:
```diff
@@ -106,6 +106,8 @@ type S3BucketSpec struct {
 	AWS AWS `json:"aws,omitempty"`
```
need a copy of line 97 please

Krenair commented on components/service-operator/apis/storage/v1beta1/s3_types_test.go:
```diff
@@ -49,6 +50,32 @@ var _ = Describe("S3Bucket", func() {
 		Expect(o.GetSecretName()).To(Equal("my-target-secret"))
 	})
 
+	It("should base egress whitelisted host name off object name", func() {
```
update this text

philandstuff commented on components/service-operator/apis/storage/v1beta1/s3_types_test.go:
```diff
@@ -49,6 +50,32 @@ var _ = Describe("S3Bucket", func() {
 		Expect(o.GetSecretName()).To(Equal("my-target-secret"))
 	})
```
copy of this test for `serviceEntry` please

philandstuff commented on components/service-operator/apis/database/v1beta1/postgres_types.go:
```diff
@@ -69,6 +70,8 @@ type PostgresSpec struct {
 	AWS PostgresAWSSpec `json:"aws,omitempty"`
 	// Secret name to be used for storing relevant instance secrets for further use.
 	Secret string `json:"secret,omitempty"`
```
copy of line 57 please

philandstuff commented on components/service-operator/apis/database/v1beta1/postgres_types_test.go:
```diff
@@ -47,6 +48,49 @@ var _ = Describe("Postgres", func() {
 		Expect(postgres.GetSecretName()).To(Equal("my-target-secret"))
```
copy of this test please

Krenair commented on components/service-operator/apis/database/v1beta1/postgres_types_test.go:
```diff
@@ -47,6 +48,49 @@ var _ = Describe("Postgres", func() {
 		Expect(postgres.GetSecretName()).To(Equal("my-target-secret"))
 	})
 
+	It("should base egress whitelisted host name off object name", func() {
```
should update this text

philandstuff reviewed (commented):
looks great. only partly reviewed but here are some comments

Krenair committed:
Address Phil's comments

philandstuff commented:
Thoughts on tests:

- can we mark cloudformation tests as skipped in short mode? (this is a go testing feature, but I don't know if ginkgo exposes it) This would mean we could run unit tests quickly in a local dev feedback loop and then run full tests before commit/PR.
- can we move the `SetupControllerEnv()` from `BeforeEach()` to `BeforeAll()` or will that break everything? (and similarly `AfterEach()` -> `AfterAll()`)  This should make things quite a bit faster.
